File: src/core/APIs/Workers/WCSharedWorker.swift
//
//  WCSharedWorker.swift
//  WTVJavaScriptCore
//
//  Created by Jing KE on 04/07/16.
//  Copyright © 2016 WizTiVi. All rights reserved.
//

import Foundation
import JavaScriptCore

class WCSharedWorker: SharedWorker {
    
    override func initMessagePort(withContext context: ScriptContext, withScope scope: SharedWorkerGlobalScope) -> MessagePort {
        let outsidePort = MessagePort.create(context)
        let insidePort = WCMessagePort.create(scope)
        MessageChannel.createChannel(withPort1: insidePort, withPort2: outsidePort)
        return outsidePort
    }
    
}

extension WCSharedWorker {
    
    override class func create(context: ScriptContext, scriptURL: String, name: String = "", options: JSValue? = nil) -> WCSharedWorker {
        if let options = options {
            return WCSharedWorker(context: context, scriptURL: scriptURL, name: name, options: WorkerOptions.create(options))
        } else {
            return WCSharedWorker(context: context, scriptURL: scriptURL)
        }
    }
    
}

File: testapps/SmartWatchWorkerExample/platforms/ios/SmartWatchWorkerExample/Resources/WCSharedWorker.js
importScripts("SharedWorker");

class WCSharedWorker extends SharedWorker {

    constructor(scriptURL, name, options, instance) {
        super(scriptURL, name, options, instance);
        if (this.constructor.name !== "WCSharedWorker") { return; }
        this.instance = instance || scope && scope.createWCSharedWorker(scriptURL, name, options);
    }

    static create(instance) {
        return new WCSharedWorker(instance.initializer.scriptURL, instance.initializer.name, null, instance);
    }

}